File: Leonidas/Final_Project/files/bbs_random.py
# crypto_project/bbs_random.py

from .bootstrap import Bootstrap
from .crypto_utils import mod_pow


class BBSRandom:
    """
    Blum-Blum-Shub random number generator.
    
    """

    def __init__(self, bit_length=512, min_value=None, max_value=None):
        self.n = None
        self.current_s = None
        self.bit_length = None

        bootstrap = Bootstrap()

        if min_value is None or max_value is None:
            # Constructor with bit_length
            self.bit_length = bit_length
            p = bootstrap.generate_prime_congruent_3_mod_4(bit_length)
            q = bootstrap.generate_prime_congruent_3_mod_4(bit_length)
            while p == q:
                q = bootstrap.generate_prime_congruent_3_mod_4(bit_length)
            self.n = p * q
            self.current_s = bootstrap.generate_seed(self.n)
        else:
            # Constructor with [min, max]
            p = bootstrap.generate_prime_in_range_congruent_3_mod_4(min_value, max_value)
            q = bootstrap.generate_prime_in_range_congruent_3_mod_4(min_value, max_value)
            while p == q:
                q = bootstrap.generate_prime_in_range_congruent_3_mod_4(min_value, max_value)
            self.n = p * q
            self.current_s = bootstrap.generate_seed(self.n)

            temp = self.n
            bits_in_n = 0
            while temp > 0:
                bits_in_n += 1
                temp >>= 1
            self.bit_length = bits_in_n

        if self.bit_length is None:
            # Fallback, falls irgendwas schief ging
            self.bit_length = 512

    def rand(self):
        """
        Generates a random number with approximately bit_length bits.
        """
        rand_number = 0
        for x in range(self.bit_length):
            self.current_s = (self.current_s * self.current_s) % self.n
            if self.current_s % 2 == 1:
                rand_number |= (1 << x)
        return rand_number

File: Alexan/final_project/src/BBSRandom.cpp
#include "BBSRandom.h"
#include "Bootstrap.h"
#include "crypto_utils.h"

BBSRandom::BBSRandom(size_t bit_length)
{
    bitLength = bit_length;

    Bootstrap b;
    const mpz_class p = b.generatePrimeCongruent3Mod4(bitLength);
    mpz_class q = b.generatePrimeCongruent3Mod4(bitLength);

    while (p == q) {
        q = b.generatePrimeCongruent3Mod4(bitLength);
    }

    n = p * q;
    current_s = b.generateSeed(n);
}

BBSRandom::BBSRandom(const mpz_class& min, const mpz_class& max)
{
    Bootstrap b;
    const mpz_class p = b.generatePrimeInRangeCongruent3Mod4(min, max);    
    mpz_class q = b.generatePrimeInRangeCongruent3Mod4(min, max);

    while (p == q) {
        q = b.generatePrimeInRangeCongruent3Mod4(min, max);
    }
    
    n = p * q;
    current_s = b.generateSeed(n);
    
    // Calculate bitLength from n
    size_t bits_in_n = 0;
    mpz_class temp = n;
    while (temp > 0) {
        bits_in_n++;
        temp = temp >> 1;
    }
    bitLength = bits_in_n;
}

BBSRandom::~BBSRandom() {}

mpz_class BBSRandom::rand()
{
    mpz_class randNumber = 0;

    for (int x = 0; x < bitLength; ++x) {
        current_s = (current_s * current_s) % n;

        if (current_s % 2 == 1) {
            mpz_setbit(randNumber.get_mpz_t(), x);
        }
    }

    return randNumber;
}